language/diem-framework/modules/CRSN.move

```diff
@@ -0,0 +1,310 @@
+        assert(size > 0, Errors::invalid_argument(EZERO_SIZE_CRSN));
+        move_to(account, CRSN {
+            min_nonce,
+            slots: BitVector::new(size),
+        })
+    }
+
+    /// Record `sequence_nonce` under the `account`. Returns true if
+    /// `sequence_nonce` is accepted, returns false if the `sequence_nonce` is rejected.
+    public(friend) fun record(account: &signer, sequence_nonce: u64): bool
+    acquires CRSN {
+        if (check(account, sequence_nonce)) {
+            // CRSN exists by `check`.
+            let crsn = borrow_global_mut<CRSN>(Signer::address_of(account));
+            // accept nonce
+            let scaled_nonce = sequence_nonce - crsn.min_nonce;
+            BitVector::set(&mut crsn.slots, scaled_nonce);
+
+            // shift the window
+            shift_window(crsn);
+
+            true
+        } else {
+            false
+        }
+    }
+
+    /// A stateless version of `record`: returns `true` if the `sequence_nonce`
+    /// will be accepted, and `false` otherwise.
+    public(friend) fun check(account: &signer, sequence_nonce: u64): bool
+    acquires CRSN {
+        let addr = Signer::address_of(account);
+        assert(has_crsn(addr), Errors::invalid_state(ENO_CRSN));
+        let crsn = borrow_global_mut<CRSN>(addr);
+
+        // Don't accept if it's outside of the window
+        if ((sequence_nonce < crsn.min_nonce) ||
+            (sequence_nonce >= crsn.min_nonce + BitVector::length(&crsn.slots))) {
+            return false
+        };
+
+        // scaled nonce is the index in the window
+        let scaled_nonce = sequence_nonce - crsn.min_nonce;
+
+        // Bit already set, reject
+        if (BitVector::is_index_set(&crsn.slots, scaled_nonce)) return false;
+
+        // otherwise, accept
+        true
+
+    }
+
+    /// Force expire transactions by forcibly shifting the window by
+    /// `shift_amount`. After the window has been shifted by `shift_amount` it is
+    /// then shifted over set bits as define by the `shift_window` function.
+    public fun force_expire(account: &signer, shift_amount: u64)
+    acquires CRSN {
+        if (shift_amount == 0) return;
+        let addr = Signer::address_of(account);
+        assert(has_crsn(addr), Errors::invalid_state(ENO_CRSN));
+        let crsn = borrow_global_mut<CRSN>(addr);
+
+        BitVector::shift_left(&mut crsn.slots, shift_amount);
+
+        crsn.min_nonce = crsn.min_nonce + shift_amount;
+        // shift over any set bits
+        shift_window(crsn);
+    }
+
+    /// Return whether this address has a CRSN resource published under it.
+    public fun has_crsn(addr: address): bool {
+        exists<CRSN>(addr)
+    }
+
+    fun shift_window(crsn: &mut CRSN) {
+        let index = 0;
+        let length = BitVector::length(&crsn.slots);
+
+        // Find the greatest index in the vector such that all indices less than it are set.
+        while (index < length) {
```

Comment:
We could consider adding this as a function in `BitVector` in case others want it + potentially to avoid some function call overhead.

Reply:
Not a bad idea! I'll add it in :)